hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/keygen/BuiltinKeyGenerator.java

```diff
@@ -238,20 +238,39 @@ private <S> S combineCompositeRecordKeyInternal(
       Predicate<S> isNullOrEmptyKeyPartPredicate,
       Object... recordKeyParts
   ) {
-    boolean hasNonNullNonEmptyPart = false;
+    if (recordKeyParts.length == 0) {
+      throw new HoodieKeyException(String.format("All of the values for (%s) were either null or empty", recordKeyFields));
+    }
 
     PartitionPathFormatterBase.StringBuilder<S> sb = builderFactory.get();
-    for (int i = 0; i < recordKeyParts.length; ++i) {
+
+    if (recordKeyParts.length == 1) {
       // NOTE: If record-key part has already been a string [[toString]] will be a no-op
-      S convertedKeyPart = emptyKeyPartHandler.apply(converter.apply(recordKeyParts[i]));
+      S convertedKeyPart = emptyKeyPartHandler.apply(converter.apply(recordKeyParts[0]));
 
-      if (encodeSingleKeyFieldName || recordKeyParts.length > 1) {
-        sb.appendJava(recordKeyFields.get(i));
+      if (encodeSingleKeyFieldName) {
+        sb.appendJava(recordKeyFields.get(0));
         sb.appendJava(DEFAULT_COLUMN_VALUE_SEPARATOR);
       }
       sb.append(convertedKeyPart);
       // This check is to validate that overall composite-key has at least one non-null, non-empty
       // segment
+      if (isNullOrEmptyKeyPartPredicate.test(convertedKeyPart)) {
+        throw new HoodieKeyException(String.format("All of the values for (%s) were either null or empty", recordKeyFields));
+      }
+      return sb.build();
+    }
+
+    boolean hasNonNullNonEmptyPart = false;
+    for (int i = 0; i < recordKeyParts.length; ++i) {
+      // NOTE: If record-key part has already been a string [[toString]] will be a no-op
+      S convertedKeyPart = emptyKeyPartHandler.apply(converter.apply(recordKeyParts[i]));
+
+      sb.appendJava(recordKeyFields.get(i));
+      sb.appendJava(DEFAULT_COLUMN_VALUE_SEPARATOR);
+      sb.append(convertedKeyPart);
+      // This check is to validate that overall composite-key has at least one non-null, non-empty
+      // segment
```

Comment:
Can this logic be simplified with just the change of `if (encodeSingleKeyFieldName || recordKeyParts.length > 1) {`?

Reply:
@danny0405 Can we follow up in a separate pr, i feel like we should not introduce any new changes here unless necessary.

Reply:
The new introduced flag is just for single record key and the logic become more clear? Also the tests should cover all these logic branches already.